src/pages/lessons/components/LessonList.tsx

```diff
@@ -108,11 +108,30 @@ function filterLessons(searchTerm: string, lessonIndex: LessonIndexEntry[]) {
   return filteredLessons;
 }
 
+function debounce<T extends Function>(cb: T, wait = 20) {
+  let h = 0;
+  let callable = (...args: any) => {
+    clearTimeout(h);
+    h = window.setTimeout(() => cb(...args), wait);
+  };
+  return callable;
+}
+
 export default function LessonList({ lessonIndex, url }: LessonListProps) {
   const location = useLocation();
-  const searchFilter = new URLSearchParams(location.search).get("q") ?? "";
+  const [searchFilter, setSearchFilter] = useState(() => new URLSearchParams(location.search).get("q") ?? "");
   const filteredLessonIndex = filterLessons(searchFilter, lessonIndex);
 
+  const history = useHistory();
+  const updateSearchParams = useMemo(() =>
+    debounce((q: string) => {
+      history.replace({ search: `?q=${q}` });
+    }, 100), [history]);
```

Comment:
Initial commit had an input lag issue: for example, when type with steno like sill -> syllable, 3 deletes and 7 adds were more visible character by character. It's because each keydown additionally required updating URL and parsing it before input can be rendered with updated value.

Now input state is back at the component and URL update is debounced and happens when no input happens after 100ms